nsiqcppstyle_types.py

```diff
@@ -1,69 +1,45 @@
 # Copyright (c) 2022 All rights reserved.
-# Redistribution and use in source and binary forms, with or without
-# modification, are permitted provided that the following conditions are
-# met:
-#
-#    * Redistributions of source code must retain the above copyright
-# notice, this list of conditions and the following disclaimer.
-#    * Redistributions in binary form must reproduce the above
-# copyright notice, this list of conditions and the following disclaimer
-# in the documentation and/or other materials provided with the
-# distribution.
-#
-# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
-# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
-# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
-# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
-# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
-# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
-# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
-# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
-# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
-# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
-# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+# SPDX-License-Identifier: BSD-3-Clause
 
 from typing import NewType, Union
 import nsiqcppstyle_lexer
 import nsiqcppstyle_checker
 
 # A stack showing the lexical context the token resides in
-ContextStackType = NewType('ContextStackType', nsiqcppstyle_checker.ContextStack)
+ContextStack = NewType('ContextStack', nsiqcppstyle_checker.ContextStack)
 
 # The new context entry that will be pushed to the contextStack due to this token
-ContextType = NewType('ContextType', nsiqcppstyle_checker.Context)
+Context = NewType('Context', nsiqcppstyle_checker.Context)
```

Comment:
For these 2, I don't think there's a need to create NewType. Just a simple `from checker import Context, ContextStack` would suffice to play the same role.

Reply:
The advantage of creating `NewType`s for `Context` and `ContextStack` is that all a script file needs to do to access hinting data types is "import nsiqcppstyle_types".  That's what is currently in nsiqcppstyle_rulemanager.py:

```
from typing import Callable
from nsiqcppstyle_types import *
```

If we remove the `NewTypes`s as you suggest, the Python script that requires hinting `Context` and `ContextStack` needs to add another import:

```
from typing import Callable
from nsiqcppstyle_types import *
from nsiqcppstyle_checker import Context, ContextStack
```

It seems somewhat non-encapsulating that, to access hinting types, `import nsiqcppstyle_types` is enough **unless** you need `Context` and `ContextStack`, in which case you need to import an additional file.  The "import once serves all" way seems to be simpler to me.

I have no problem either way, let's just finalize the decision.

Reply:
You can re-expose the types from the _types file so the users don't need to import the _checker

Reply:
`nsiqcppstyle_types.py` now looks like this:

```
from typing import NewType, Union
import nsiqcppstyle_lexer
import nsiqcppstyle_checker
from nsiqcppstyle_checker import Context, ContextStack
```

Because Context and ContextStack are not used in _types.py itself, the IDE (and Clang-Tidy) show a warning for the fourth line: "Unused import statement 'from nsiqcppstyle_checker import Context, ContextStack'".  These are used, of course, in user rule files.

Done.

Reply:
What if you did the following instead:
```py
import nsiqcppstyle_checker

Context = nsiqcppstyle_checker.Context
```

Would the IDE still complain? AFAIK, this will not change the semantics

Reply:
As per your comment, I modified the code in nsiqcppstyle_types.py to the following:

<pre><code>
from typing import NewType, Union
import nsiqcppstyle_lexer
import nsiqcppstyle_checker
<del>from nsiqcppstyle_checker import Context, ContextStack</del>

# Declare already existing types (i.e., Context and ContextStack)
Context = nsiqcppstyle_checker.Context
ContextStack = nsiqcppstyle_checker.ContextStack
</code></pre>

To use `Context` or `ContextStack` in a user rule:

```
from nsiqunittest.nsiqcppstyle_unittestbase import *
from nsiqcppstyle_rulehelper import *
from nsiqcppstyle_reporter import *
from nsiqcppstyle_rulemanager import *
from nsiqcppstyle_types import *

def PreprocessRule(lexer: Lexer, contextStack: ContextStack):
    pass
...
ruleManager.AddPreprocessRule(PreprocessRule)
...
```

The IDE no longer emits a warning.

Done.